doom_emacs/.config/doom/config.org
:PROPERTIES:
:ID:       a7514bf6-07b6-45ab-9838-db50dcc75526
:END:
#+title:    Main configuration
#+subtitle: ;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

* Introduction to doom
 Place your private configuration here! Remember, you do not need to run 'doom
 sync' after modifying this file!


 Some functionality uses this to identify you, e.g. GPG configuration, email
 clients, file templates and snippets. It is optional.
 (setq user-full-name "John Doe"
       user-mail-address "john@doe.com")

 Doom exposes five (optional) variables for controlling fonts in Doom:

 - `doom-font' -- the primary font to use
 - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
 - `doom-big-font' -- used for `doom-big-font-mode'; use this for
   presentations or streaming.
 - `doom-symbol-font' -- for symbols
 - `doom-serif-font' -- for the `fixed-pitch-serif' face

 See 'C-h v doom-font' for documentation and more examples of what they
 accept. For example:

(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))

 If you or Emacs can't find your font, use 'M-x describe-font' to look them
 up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
 refresh your font settings. If Emacs still can't find your font, it likely
 wasn't installed correctly. Font issues are rarely Doom issues!

** Auth service
#+begin_src emacs-lisp
(require 'auth-source)
(setq auth-sources '("~/.authinfo.gpg"))
#+end_src

** Integrate omarchy themes
#+begin_src emacs-lisp
(load! "omarchy-themes")
(setq doom-theme 'doom-old-hope)
#+end_src

 There are two ways to load a theme. Both assume the theme is installed and
 available. You can either set `doom-theme' or manually load a theme with the
 `load-theme' function. This is the default:
 (setq doom-theme 'doom-1337)

** Line numbers

 This determines the style of line numbers in effect. If set to `nil', line
 numbers are disabled. For relative line numbers, set this to `relative'.
 #+begin_src emacs-lisp
(setq display-line-numbers-type 'relative)
 #+end_src

** Key-bindings
#+begin_src emacs-lisp
(after! centaur-tabs
  (evil-global-set-key 'normal "H" #'centaur-tabs-backward-tab)
  (evil-global-set-key 'normal "L" #'centaur-tabs-forward-tab)
  (evil-global-set-key 'normal "to" #'centaur-tabs-toggle-groups)
  )
#+end_src

** Org directory
 If you use `org' and don't want your org files in the default location below,
 change `org-directory'. It must be set before org loads!
#+begin_src emacs-lisp
(setq org-directory "~/org/")
(setq org-roam-directory "~/org")
#+end_src

** Adding verb to org
#+begin_src emacs-lisp
(use-package org
:mode ("\\.org\\'" . org-mode)
:config (define-key org-mode-map (kbd "C-c C-r") verb-command-map))
#+end_src

** Org roam ui
#+begin_src emacs-lisp
(use-package! websocket
    :after org-roam)

(use-package! org-roam-ui
    :after org-roam ;; or :after org
;;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
;;         a hookable mode anymore, you're advised to pick something yourself
;;         if you don't care about startup time, use
;;  :hook (after-init . org-roam-ui-mode)
    :config
    (setq org-roam-ui-sync-theme t
          org-roam-ui-follow t
          org-roam-ui-update-on-save t
          org-roam-ui-open-on-start t))
#+end_src

** Font
#+begin_src emacs-lisp
(setq doom-font
      (font-spec :family "Liberation Mono" :size 10)
      doom-variable-pitch-font
      (font-spec :family "Liberation Sans" :size 11))

(after! org
  (set-face-attribute 'org-level-1 nil :height 1.5)
  (set-face-attribute 'org-level-2 nil :height 1.3)
  (set-face-attribute 'org-level-3 nil :height 1.15)
  (set-face-attribute 'org-level-4 nil :height 1.05))

(after! markdown-mode
  (set-face-attribute 'markdown-header-face-1 nil :height 1.5)
  (set-face-attribute 'markdown-header-face-2 nil :height 1.3)
  (set-face-attribute 'markdown-header-face-3 nil :height 1.15)
  (set-face-attribute 'markdown-header-face-4 nil :height 1.05))
#+end_src

** Escape sec for evil
#+begin_src emacs-lisp
(after! evil
  (setq evil-escape-key-sequence "jk")
  (setq evil-escape-delay 0.2))
#+end_src

** Eglot
#+begin_src emacs-lisp
(require 'eglotx-eglot)

(after! eglot

  ;; TSX / JSX: VTSLS + Tailwind
  (add-to-list
   'eglot-server-programs
   `(((tsx-ts-mode :language-id "typescriptreact")
      (js-mode :language-id "javascriptreact")
      (js-ts-mode :language-id "javascriptreact"))
     . ,(eglotx-contact
         '(:name "vtsls"
           :command ("/home/blackdovah/.local/share/nvim/mason/bin/vtsls"
                     "--stdio")
           :priority 100
           :required t)
         '(:name "tailwindcss"
           :command ("/home/blackdovah/.local/share/nvim/mason/bin/tailwindcss-language-server"
                     "--stdio")
           :priority 120
           :required nil))))

  ;; TypeScript: VTSLS only
  (add-to-list
   'eglot-server-programs
   `(((typescript-ts-mode :language-id "typescript")
      (typescript-mode :language-id "typescript"))
     . ("/home/blackdovah/.local/share/nvim/mason/bin/vtsls"
        "--stdio")))

  ;; JavaScript: VTSLS only
  (add-to-list
   'eglot-server-programs
   `(((js-mode :language-id "javascript")
      (js-ts-mode :language-id "javascript"))
     . ("/home/blackdovah/.local/share/nvim/mason/bin/vtsls"
        "--stdio"))))

(setq-default
 eglot-workspace-configuration
 '(:vtsls
   (:enableMoveToFileCodeAction t
    :autoUseWorkspaceTsdk t
    :experimental
    (:completion
     (:enableServerSideFuzzyMatch t)))

   :typescript
   (:updateImportsOnFileMove
    (:enabled "always")
    :suggest
    (:completeFunctionCalls t)
    (:enumMemberValues (:enabled t)
     :functionLikeReturnTypes (:enabled t)
     :parameterNames (:enabled "literals")
     :parameterTypes (:enabled t)
     :propertyDeclarationTypes (:enabled t)
     :variableTypes (:enabled t)))

   :javascript
   (:suggest
    (:completeFunctionCalls t))))
#+end_src

** Eglotx
#+begin_src emacs-lisp
(after! eglotx
  (eglotx-presets-mode 1))
#+end_src

** Corfu tailwind completions fix
- Emacs considers '-' to be punctuation, and corfu treated it as a separator
- That broke tailwind's completions as they rely heavily on '-'
- This fix removes the quit boundary (default = separator) whenever tsx-ts-mode is on
#+begin_src emacs-lisp
(after! corfu
  (add-hook 'tsx-ts-mode-hook
            (lambda ()
              (setq-local corfu-quit-at-boundary nil))))
#+end_src
** Set SSH agent via keychain
#+begin_src emacs-lisp
(let ((auth-sock
       (string-trim
        (shell-command-to-string
         "keychain --eval --quiet | sed -n 's/^SSH_AUTH_SOCK=\\([^;]*\\).*/\\1/p'"))))
  (unless (string-empty-p auth-sock)
    (setenv "SSH_AUTH_SOCK" auth-sock)))
#+end_src

** Clutch config for DB2 for i
#+begin_src emacs-lisp
(setq clutch-jdbc-agent-java-executable
      "/home/blackdovah/.local/share/mise/installs/java/liberica-26.0.0+37/bin/java")
(setq clutch-connection-alist
      `(("va-ibmi" .(:backend jdbc
                     :url "jdbc:as400://10.30.1.134;keep alive=true;metadata source=0;"
                     :driver-class "com.ibm.as400.access.AS400JDBCDriver"
                     :user "esky001"
                     :password ,(auth-source-pick-first-password
                                 :host "va-ibmi"
                                 :user "esky001")))
        ;; Config for redis
        ("va-redis" . (:backend redis
                       :host "localhost"
                       :port 6379
                       :password ,(auth-source-pick-first-password
                                 :host "va-redis")))
        ("PUB400" .(:backend jdbc
                    :url "jdbc:as400://PUB400.COM;keep alive=true;metadata source=0;"
                    :driver-class "com.ibm.as400.access.AS400JDBCDriver"
                    :user "BLACKDOVA"
                    :password ,(auth-source-pick-first-password
                                 :host "PUB400"
                                 :user "BLACKDOVA")))
        ))
#+end_src

** tmux-pane config
#+begin_src emacs-lisp
(use-package! tmux-pane
  :config
  (tmux-pane-mode)
  (map! :leader
        (:prefix ("v" . "tmux pane")
         :desc "Open vpane" :nv "o" #'tmux-pane-open-vertical
         :desc "Open hpane" :nv "h" #'tmux-pane-open-horizontal
         :desc "Open hpane" :nv "s" #'tmux-pane-open-horizontal
         :desc "Open vpane" :nv "v" #'tmux-pane-open-vertical
         :desc "Close pane" :nv "c" #'tmux-pane-close
         :desc "Rerun last command" :nv "r" #'tmux-pane-rerun))
  (map! :leader
        (:prefix "t"
         :desc "vpane" :nv "v" #'tmux-pane-toggle-vertical
         :desc "hpane" :nv "h" #'tmux-pane-toggle-horizontal)))
#+end_src

** Banner
#+begin_src emacs-lisp
(defun +dashboard-draw-ascii-banner-fn ()
  (propertize
   (with-temp-buffer
     (insert-file-contents
      (expand-file-name "banner.txt" doom-user-dir))
     (buffer-string))
   'face '+dashboard-banner))
#+end_src

** org-mode better bullets
#+begin_src emacs-lisp
(add-hook 'org-mode-hook 'org-indent-mode)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+end_src

** Temporary projectile fix until doom introduces a fix
#+begin_src emacs-lisp
(after! projectile
  (global-unset-key [remap evil-jump-to-tag])
  (global-unset-key [remap find-tag]))
#+end_src

** org-auto-tangle
#+begin_src emacs-lisp
(use-package! org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config
  (setq org-auto-tangle-default t))
#+end_src

** Rstudio mode
#+begin_src emacs-lisp
;;; R / ESS

(after! ess-r-mode

  ;; RStudio-compatible indentation
  (setq ess-style 'RStudio)

  ;; Don't ask where to start R
  (setq ess-ask-for-ess-directory nil)

  ;; Don't restore .RData
  (setq inferior-R-args "--no-save")

  ;; Normal comment indentation
  (setq ess-indent-with-fancy-comments nil)

  ;; Don't block Emacs during evaluation
  (setq ess-eval-visibly 'nowait)

  (defun my/rstudio-layout ()
    "Build an RStudio-like layout."
    (interactive)

    (let ((source (current-buffer)))

      ;; Start R
      (ess-switch-to-ESS t)

      ;; Open the environment
      (ess-rdired)

      ;; Get the actual ESS buffers
      (let ((console (ess-get-current-process-buffer))
            (environment (get-buffer ess-rdired-buffer)))

        ;; Go back to our R source file
        (pop-to-buffer source)
        (delete-other-windows)

        ;; Console -> right of source
        (display-buffer-in-atom-window
         console
         `((window . ,(get-buffer-window source))
           (side . right)
           (window-width . 0.33)
           (dedicated . t)))

        ;; Environment -> below console
        (display-buffer-in-atom-window
         environment
         `((window . ,(get-buffer-window console))
           (side . below)
           (window-height . 0.35)
           (dedicated . t))))



      ;; (display-buffer-in-atom-window
      ;;  plots
      ;;  `((window . ,(get-buffer-window source))
      ;;   (side . below) (window-height . 0.35)
      ;;   (dedicated . t))))
      ))
  ;; (add-hook 'ess-r-mode-hook #'my/rstudio-layout)
  )
#+end_src

** babel dispatcher
#+begin_src emacs-lisp
(require 'ob-lob)

(org-babel-lob-ingest "~/stow_dotfiles/doom_emacs/.config/doom/functions.org")
(load! "lisp/custom-functions")
(keymap-global-set "C-x w w" 'my-babel-functions)
#+end_src

** gptel
#+begin_src emacs-lisp
(gptel-make-ollama
    "Ollama"
  :host "localhost:11434"
  :stream t
  :models '(
            qwen3:8b-40960ctx
            qwen3:4b-16384ctx
            qwen3:8b-16384ctx
            qwen2.5-coder:7b-16384ctx
            qwen2.5-coder:7b
            qwen3:8b
            llama3:8b
            hf.co/unsloth/Qwen3-4B-GGUF:Q4_K_M
            ))

(gptel-make-openai
 "OpenRouter"
 :host "openrouter.ai"
 :endpoint "/api/v1/chat/completions"
 :stream t
 :key (lambda ()
 (auth-source-pick-first-password
 :host "openrouter.ai"
 :user "apikey"))
 :models '(openrouter/free))
#+end_src

** gptel-agent
#+begin_src emacs-lisp
(after! gptel-agent
  (gptel-agent-update))
#+end_src

** Footer
 Whenever you reconfigure a package, make sure to wrap your config in an
 `with-eval-after-load' block, otherwise Doom's defaults may override your
 settings. E.g.

   (with-eval-after-load 'PACKAGE
     (setq x y))

 The exceptions to this rule:

   - Setting file/directory variables (like `org-directory')
   - Setting variables which explicitly tell you to set them before their
     package is loaded (see 'C-h v VARIABLE' to look them up).
   - Setting doom variables (which start with 'doom-' or '+').

 Here are some additional functions/macros that will help you configure Doom.

 - `load!' for loading external *.el files relative to this one
 - `add-load-path!' for adding directories to the `load-path', relative to
   this file. Emacs searches the `load-path' when you load packages with
   `require' or `use-package'.
 - `map!' for binding new keys

 To get information about any of these functions/macros, move the cursor over
 the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
 This will open documentation for it, including demos of how they are used.
 Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
 etc).

 You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
 they are implemented.


doom_emacs/.config/doom/config.el
(require 'auth-source)
(setq auth-sources '("~/.authinfo.gpg"))

(load! "omarchy-themes")
(setq doom-theme 'doom-old-hope)

(setq display-line-numbers-type 'relative)

(after! centaur-tabs
  (evil-global-set-key 'normal "H" #'centaur-tabs-backward-tab)
  (evil-global-set-key 'normal "L" #'centaur-tabs-forward-tab)
  (evil-global-set-key 'normal "to" #'centaur-tabs-toggle-groups)
  )

(setq org-directory "~/org/")
(setq org-roam-directory "~/org")

(use-package org
:mode ("\\.org\\'" . org-mode)
:config (define-key org-mode-map (kbd "C-c C-r") verb-command-map))

(use-package! websocket
    :after org-roam)

(use-package! org-roam-ui
    :after org-roam ;; or :after org
;;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
;;         a hookable mode anymore, you're advised to pick something yourself
;;         if you don't care about startup time, use
;;  :hook (after-init . org-roam-ui-mode)
    :config
    (setq org-roam-ui-sync-theme t
          org-roam-ui-follow t
          org-roam-ui-update-on-save t
          org-roam-ui-open-on-start t))

(setq doom-font
      (font-spec :family "Liberation Mono" :size 10)
      doom-variable-pitch-font
      (font-spec :family "Liberation Sans" :size 11))

(after! org
  (set-face-attribute 'org-level-1 nil :height 1.5)
  (set-face-attribute 'org-level-2 nil :height 1.3)
  (set-face-attribute 'org-level-3 nil :height 1.15)
  (set-face-attribute 'org-level-4 nil :height 1.05))

(after! markdown-mode
  (set-face-attribute 'markdown-header-face-1 nil :height 1.5)
  (set-face-attribute 'markdown-header-face-2 nil :height 1.3)
  (set-face-attribute 'markdown-header-face-3 nil :height 1.15)
  (set-face-attribute 'markdown-header-face-4 nil :height 1.05))

(after! evil
  (setq evil-escape-key-sequence "jk")
  (setq evil-escape-delay 0.2))

(require 'eglotx-eglot)

(after! eglot

  ;; TSX / JSX: VTSLS + Tailwind
  (add-to-list
   'eglot-server-programs
   `(((tsx-ts-mode :language-id "typescriptreact")
      (js-mode :language-id "javascriptreact")
      (js-ts-mode :language-id "javascriptreact"))
     . ,(eglotx-contact
         '(:name "vtsls"
           :command ("/home/blackdovah/.local/share/nvim/mason/bin/vtsls"
                     "--stdio")
           :priority 100
           :required t)
         '(:name "tailwindcss"
           :command ("/home/blackdovah/.local/share/nvim/mason/bin/tailwindcss-language-server"
                     "--stdio")
           :priority 120
           :required nil))))

  ;; TypeScript: VTSLS only
  (add-to-list
   'eglot-server-programs
   `(((typescript-ts-mode :language-id "typescript")
      (typescript-mode :language-id "typescript"))
     . ("/home/blackdovah/.local/share/nvim/mason/bin/vtsls"
        "--stdio")))

  ;; JavaScript: VTSLS only
  (add-to-list
   'eglot-server-programs
   `(((js-mode :language-id "javascript")
      (js-ts-mode :language-id "javascript"))
     . ("/home/blackdovah/.local/share/nvim/mason/bin/vtsls"
        "--stdio"))))

(setq-default
 eglot-workspace-configuration
 '(:vtsls
   (:enableMoveToFileCodeAction t
    :autoUseWorkspaceTsdk t
    :experimental
    (:completion
     (:enableServerSideFuzzyMatch t)))

   :typescript
   (:updateImportsOnFileMove
    (:enabled "always")
    :suggest
    (:completeFunctionCalls t)
    (:enumMemberValues (:enabled t)
     :functionLikeReturnTypes (:enabled t)
     :parameterNames (:enabled "literals")
     :parameterTypes (:enabled t)
     :propertyDeclarationTypes (:enabled t)
     :variableTypes (:enabled t)))

   :javascript
   (:suggest
    (:completeFunctionCalls t))))

(after! eglotx
  (eglotx-presets-mode 1))

(after! corfu
  (add-hook 'tsx-ts-mode-hook
            (lambda ()
              (setq-local corfu-quit-at-boundary nil))))

(let ((auth-sock
       (string-trim
        (shell-command-to-string
         "keychain --eval --quiet | sed -n 's/^SSH_AUTH_SOCK=\\([^;]*\\).*/\\1/p'"))))
  (unless (string-empty-p auth-sock)
    (setenv "SSH_AUTH_SOCK" auth-sock)))

(setq clutch-jdbc-agent-java-executable
      "/home/blackdovah/.local/share/mise/installs/java/liberica-26.0.0+37/bin/java")
(setq clutch-connection-alist
      `(("va-ibmi" .(:backend jdbc
                     :url "jdbc:as400://10.30.1.134;keep alive=true;metadata source=0;"
                     :driver-class "com.ibm.as400.access.AS400JDBCDriver"
                     :user "esky001"
                     :password ,(auth-source-pick-first-password
                                 :host "va-ibmi"
                                 :user "esky001")))
        ;; Config for redis
        ("va-redis" . (:backend redis
                       :host "localhost"
                       :port 6379
                       :password ,(auth-source-pick-first-password
                                 :host "va-redis")))
        ("PUB400" .(:backend jdbc
                    :url "jdbc:as400://PUB400.COM;keep alive=true;metadata source=0;"
                    :driver-class "com.ibm.as400.access.AS400JDBCDriver"
                    :user "BLACKDOVA"
                    :password ,(auth-source-pick-first-password
                                 :host "PUB400"
                                 :user "BLACKDOVA")))
        ))

(use-package! tmux-pane
  :config
  (tmux-pane-mode)
  (map! :leader
        (:prefix ("v" . "tmux pane")
         :desc "Open vpane" :nv "o" #'tmux-pane-open-vertical
         :desc "Open hpane" :nv "h" #'tmux-pane-open-horizontal
         :desc "Open hpane" :nv "s" #'tmux-pane-open-horizontal
         :desc "Open vpane" :nv "v" #'tmux-pane-open-vertical
         :desc "Close pane" :nv "c" #'tmux-pane-close
         :desc "Rerun last command" :nv "r" #'tmux-pane-rerun))
  (map! :leader
        (:prefix "t"
         :desc "vpane" :nv "v" #'tmux-pane-toggle-vertical
         :desc "hpane" :nv "h" #'tmux-pane-toggle-horizontal)))

(defun +dashboard-draw-ascii-banner-fn ()
  (propertize
   (with-temp-buffer
     (insert-file-contents
      (expand-file-name "banner.txt" doom-user-dir))
     (buffer-string))
   'face '+dashboard-banner))

(add-hook 'org-mode-hook 'org-indent-mode)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(after! projectile
  (global-unset-key [remap evil-jump-to-tag])
  (global-unset-key [remap find-tag]))

(use-package! org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config
  (setq org-auto-tangle-default t))

;;; R / ESS

(after! ess-r-mode

  ;; RStudio-compatible indentation
  (setq ess-style 'RStudio)

  ;; Don't ask where to start R
  (setq ess-ask-for-ess-directory nil)

  ;; Don't restore .RData
  (setq inferior-R-args "--no-save")

  ;; Normal comment indentation
  (setq ess-indent-with-fancy-comments nil)

  ;; Don't block Emacs during evaluation
  (setq ess-eval-visibly 'nowait)

  (defun my/rstudio-layout ()
    "Build an RStudio-like layout."
    (interactive)

    (let ((source (current-buffer)))

      ;; Start R
      (ess-switch-to-ESS t)

      ;; Open the environment
      (ess-rdired)

      ;; Get the actual ESS buffers
      (let ((console (ess-get-current-process-buffer))
            (environment (get-buffer ess-rdired-buffer)))

        ;; Go back to our R source file
        (pop-to-buffer source)
        (delete-other-windows)

        ;; Console -> right of source
        (display-buffer-in-atom-window
         console
         `((window . ,(get-buffer-window source))
           (side . right)
           (window-width . 0.33)
           (dedicated . t)))

        ;; Environment -> below console
        (display-buffer-in-atom-window
         environment
         `((window . ,(get-buffer-window console))
           (side . below)
           (window-height . 0.35)
           (dedicated . t))))



      ;; (display-buffer-in-atom-window
      ;;  plots
      ;;  `((window . ,(get-buffer-window source))
      ;;   (side . below) (window-height . 0.35)
      ;;   (dedicated . t))))
      ))
  ;; (add-hook 'ess-r-mode-hook #'my/rstudio-layout)
  )

(require 'ob-lob)

(org-babel-lob-ingest "~/stow_dotfiles/doom_emacs/.config/doom/functions.org")
(load! "lisp/custom-functions")
(keymap-global-set "C-x w w" 'my-babel-functions)

(gptel-make-ollama
    "Ollama"
  :host "localhost:11434"
  :stream t
  :models '(
            qwen3:8b-40960ctx
            qwen3:4b-16384ctx
            qwen3:8b-16384ctx
            qwen2.5-coder:7b-16384ctx
            qwen2.5-coder:7b
            qwen3:8b
            llama3:8b
            hf.co/unsloth/Qwen3-4B-GGUF:Q4_K_M
            ))

(gptel-make-openai
 "OpenRouter"
 :host "openrouter.ai"
 :endpoint "/api/v1/chat/completions"
 :stream t
 :key (lambda ()
 (auth-source-pick-first-password
 :host "openrouter.ai"
 :user "apikey"))
 :models '(openrouter/free))

(after! gptel-agent
  (gptel-agent-update))
